torch/utils/cpp_extension.py

```diff
@@ -1177,65 +1321,132 @@ def _write_ninja_file(path,
+        flags.append('cuda_post_cflags = {}'.format(' '.join(cuda_post_cflags)))
     flags.append('ldflags = {}'.format(' '.join(ldflags)))
 
+    # Turn into absolute paths so we can emit them into the ninja build
+    # file wherever it is.
+    sources = [os.path.abspath(file) for file in sources]
+
     # See https://ninja-build.org/build.ninja.html for reference.
     compile_rule = ['rule compile']
     if IS_WINDOWS:
         compile_rule.append(
-            '  command = cl /showIncludes $cflags -c $in /Fo$out')
+            '  command = cl /showIncludes $cflags -c $in /Fo$out $post_cflags')
         compile_rule.append('  deps = msvc')
     else:
         compile_rule.append(
-            '  command = $cxx -MMD -MF $out.d $cflags -c $in -o $out')
+            '  command = $cxx -MMD -MF $out.d $cflags -c $in -o $out $post_cflags')
         compile_rule.append('  depfile = $out.d')
         compile_rule.append('  deps = gcc')
 
     if with_cuda:
         cuda_compile_rule = ['rule cuda_compile']
         cuda_compile_rule.append(
-            '  command = $nvcc $cuda_flags -c $in -o $out')
-
-    link_rule = ['rule link']
-    if IS_WINDOWS:
-        cl_paths = subprocess.check_output(['where',
-                                            'cl']).decode().split('\r\n')
-        if len(cl_paths) >= 1:
-            cl_path = os.path.dirname(cl_paths[0]).replace(':', '$:')
-        else:
-            raise RuntimeError("MSVC is required to load C++ extensions")
-        link_rule.append(
-            '  command = "{}/link.exe" $in /nologo $ldflags /out:$out'.format(
-                cl_path))
-    else:
-        link_rule.append('  command = $cxx $in $ldflags -o $out')
+            '  command = $nvcc $cuda_cflags -c $in -o $out $cuda_post_cflags')
 
     # Emit one build rule per source to enable incremental build.
-    object_files = []
     build = []
-    for source_file in sources:
-        # '/path/to/file.cpp' -> 'file'
-        file_name = os.path.splitext(os.path.basename(source_file))[0]
-        if _is_cuda_file(source_file) and with_cuda:
-            rule = 'cuda_compile'
-            # Use a different object filename in case a C++ and CUDA file have
-            # the same filename but different extension (.cpp vs. .cu).
-            target = '{}.cuda.o'.format(file_name)
-        else:
-            rule = 'compile'
-            target = '{}.o'.format(file_name)
-        object_files.append(target)
+    for source_file, object_file in zip(sources, objects):
+        is_cuda_source = _is_cuda_file(source_file) and with_cuda
+        rule = 'cuda_compile' if is_cuda_source else 'compile'
         if IS_WINDOWS:
             source_file = source_file.replace(':', '$:')
         source_file = source_file.replace(" ", "$ ")
-        build.append('build {}: {} {}'.format(target, rule, source_file))
+        build.append('build {}: {} {}'.format(object_file, rule, source_file))
```

Comment:
It looks like you're messing around with the Windows code now. Did you get it to work?

Reply:
Yeah, just-in-time cpp extensions works with ninja. I didn't get ahead-of-time cpp extensions to work with ninja, but the just-in-time and ahead-of-time variants share code. It should be possible to (with a little extra work) get windows working with ahead-of-time cpp extensions; I'm planning on filing an issue for that after this goes in